exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,262 @@
+		agg := record.Aggregator()
+		if sum, ok := agg.(aggregator.Sum); ok {
+			value, err := sum.Sum()
+			if err != nil {
+				aggError = err
+				// TODO: halp, where are we doing here?
+				return
+			}
+
+			c, err := e.getCounter(record)
+			if err != nil {
+				// TODO: log a warning here?
+				return
+			}
+
+			c.Add(value.AsFloat64())
+		}
+
+		if gauge, ok := agg.(aggregator.LastValue); ok {
+			lv, _, err := gauge.LastValue()
+			if err != nil {
+				aggError = err
+				// TODO: halp
+				return
+			}
+
+			g, err := e.getGauge(record)
+			if err != nil {
+				// TODO: log a warning here?
+				return
+			}
+			g.Set(lv.AsFloat64())
+		}
+	})
+
+	return aggError
+}
+
+func (e *Exporter) getCounter(record export.Record) (prometheus.Counter, error) {
+	e.Lock()
+	defer e.Unlock()
+
+	desc := record.Descriptor()
+	if c, ok := e.counters[desc]; ok {
+		return c, nil
+	}
+
+	counterVec, err := e.getCounterVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	counter := counterVec.With(labelsToTags(record.Labels()))
+
+	e.counters[desc] = counter
+	return counter, nil
+}
+
+func (e *Exporter) getGauge(record export.Record) (prometheus.Gauge, error) {
+	e.Lock()
+	defer e.Unlock()
+
+	desc := record.Descriptor()
+	if g, ok := e.gauges[desc]; ok {
+		return g, nil
+	}
+
+	gaugeVec, err := e.getGaugeVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	gauge := gaugeVec.With(labelsToTags(record.Labels()))
+
+	e.gauges[desc] = gauge
+	return gauge, nil
+}
+
+func (e *Exporter) getCounterVec(desc *export.Descriptor, labels export.Labels) (*prometheus.CounterVec, error) {
+	id, tagKeys := getCanonicalID(desc, labels)
+
+	if c, ok := e.counterVecs[id]; ok {
+		return c, nil
+	}
+
+	c := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: Sanitize(desc.Name()),
+			Help: desc.Description(),
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(c); err != nil {
+		return nil, err
+	}
+
+	e.counterVecs[id] = c
+	return c, nil
+}
+
+func (e *Exporter) getGaugeVec(desc *export.Descriptor, labels export.Labels) (*prometheus.GaugeVec, error) {
+	id, tagKeys := getCanonicalID(desc, labels)
+
+	if g, ok := e.gaugeVecs[id]; ok {
+		return g, nil
+	}
+
+	g := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: Sanitize(desc.Name()),
+			Help: desc.Description(),
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(g); err != nil {
+		return nil, err
+	}
+
+	e.gaugeVecs[id] = g
+	return g, nil
+}
+
+func getTagKeys(keys []core.KeyValue) []string {
+	tagKeys := make([]string, 0, len(keys))
+	for _, kv := range keys {
+		tagKeys = append(tagKeys, Sanitize(string(kv.Key)))
+	}
+	return tagKeys
+}
+
+func getCanonicalID(desc *export.Descriptor, labels export.Labels) (string, []string) {
+	tagKeys := getTagKeys(labels.Ordered())
+	return Sanitize(desc.Name()) + prefixSplitter + labels.Encoded(), tagKeys
+}
+
+func labelsToTags(labels export.Labels) map[string]string {
```

Comment:
(I wonder if Prometheus has a way to provide a slice of values in the same order the tag keys were declared on the vector? Constructing a map here seems wasteful.)

Reply:
Don't think so. Prometheus uses a `map[string]string` for labels.

Reply:
Yeah, I don't think it is available in the public API of the prometheus client.

Reply:
Actually, there is. I'll see how to use it tomorrow.

Reply:
Yeah :D, We can use `GetMetricWithLabelValues(lvs ...string)` instead of  `GetMetricWith(labels Labels)`.